eyeball-im-util/src/vector/ops.rs

```diff
@@ -1,71 +1,75 @@
-use std::collections::VecDeque;
-
+use arrayvec::ArrayVec;
 use eyeball_im::VectorDiff;
 
-pub trait VectorDiffContainerOps<T> {
+pub trait VectorDiffContainerOps<T>: Sized {
     type Family: VectorDiffContainerFamily;
+    type Buffer: Default;
 
     fn from_item(vector_diff: VectorDiff<T>) -> Self;
 
-    fn pick_item(vector_diffs: &mut VecDeque<Self>) -> Option<Self>
-    where
-        Self: Sized;
-
-    fn for_each(self, f: impl FnMut(VectorDiff<T>));
-
     fn filter_map<U>(
         self,
         f: impl FnMut(VectorDiff<T>) -> Option<VectorDiff<U>>,
     ) -> Option<VectorDiffContainerFamilyMember<Self::Family, U>>;
+
+    fn push_into_buffer(
+        self,
+        buffer: &mut Self::Buffer,
+        make_diffs: impl FnMut(VectorDiff<T>) -> ArrayVec<VectorDiff<T>, 2>,
+    ) -> Option<Self>;
+
+    fn pop_from_buffer(buffer: &mut Self::Buffer) -> Option<Self>;
 }
 
 #[allow(unreachable_pub)]
 pub type VectorDiffContainerFamilyMember<F, U> = <F as VectorDiffContainerFamily>::Member<U>;
 
 impl<T> VectorDiffContainerOps<T> for VectorDiff<T> {
     type Family = VectorDiffFamily;
+    type Buffer = Option<VectorDiff<T>>;
 
     fn from_item(vector_diff: VectorDiff<T>) -> Self {
         vector_diff
     }
 
-    fn pick_item(vector_diffs: &mut VecDeque<Self>) -> Option<Self>
-    where
-        Self: Sized,
-    {
-        vector_diffs.pop_front()
-    }
-
-    fn for_each(self, mut f: impl FnMut(VectorDiff<T>)) {
-        f(self);
-    }
-
     fn filter_map<U>(
         self,
         mut f: impl FnMut(VectorDiff<T>) -> Option<VectorDiff<U>>,
     ) -> Option<VectorDiffContainerFamilyMember<Self::Family, U>> {
         f(self)
     }
+
+    fn push_into_buffer(
+        self,
+        buffer: &mut Self::Buffer,
+        mut make_diffs: impl FnMut(VectorDiff<T>) -> ArrayVec<VectorDiff<T>, 2>,
+    ) -> Option<Self> {
+        assert!(buffer.is_none(), "limit buffer must be None when calling push_limit_buf");
```

Comment:
I'm not fine mentionning `limit` here. Maybe just: “buffer must be `None` when calling `push_limit_buf`”?